crypto/asn1/d2i_pr.c

```diff
@@ -40,8 +40,14 @@ d2i_PrivateKey_decoder(int keytype, EVP_PKEY **a, const unsigned char **pp,
         if (key_name == NULL)
             return NULL;
     }
-    if (a != NULL && *a != NULL)
+    if ((a == NULL) || (*a == NULL)) {
+        if ((*ppkey = EVP_PKEY_new()) == NULL) {
+            ERR_raise(ERR_LIB_ASN1, ERR_R_EVP_LIB);
+            return NULL;
+        }
+    } else {
```

Comment:
Why should this be needed? The OSSL_DECODER_CTX_new_for_pkey is supposed to handle the *ppkey == NULL case by allocating a new EVP_PKEY.

Reply:
Things are unfortunately subtle. Due to the wonderful legacy semantics of 
`EVP_PKEY *d2i_PrivateKey(int type, EVP_PKEY **a, ...)`
which is not properly documented,
on success these functions should not only return a non-NULL key pointer 
but also assign this pointer to `*a` in case `a` was non-NULL but `*a` was NULL.
This special feature got lost in the mentioned rather recent commit.

Reply:
Yes, but how does this change fixes this?

Reply:
I'd expect before the `return *ppkey;` below something like `if (a != NULL && ppkey != a) *a = *ppkey;`

Reply:
There is also even more subtle problem with this code - in case `*a != NULL` and `evp_keymgmt_util_has(*ppkey, OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 1` the *a is modified but NULL is returned. Not sure it is worth fixing.

Reply:
I meanwhile noticed that I fixed the function not in the right way; done now.
On this occasion also properly documented this behavior:
```
If successful and the I<a> parameter is not NULL the function assigns the
returned B<EVP_KEY> structure pointer to I<*a>, overwriting any previous value.
```
and removed a wrong old piece of documentation (was was wrong already in 1.1.1):
```
If the I<*a> is not NULL when calling d2i_PrivateKey() or d2i_AutoPrivateKey()
(i.e. an existing structure is being reused) and the key format is PKCS#8
then I<*a> will be freed and replaced on a successful call.
```

Reply:
> There is also even more subtle problem with this code - in case `*a != NULL` and `evp_keymgmt_util_has(*ppkey, OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 1` the *a is modified but NULL is returned. Not sure it is worth fixing.

Really subtle. I think I've managed to fix that.